streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java

```diff
@@ -166,12 +174,22 @@ public void closeSuspended(final boolean clean,
     public List<ConsumerRecord<byte[], byte[]>> update(final TopicPartition partition,
                                                        final List<ConsumerRecord<byte[], byte[]>> records) {
         log.trace("Updating standby replicas of its state store for partition [{}]", partition);
-        final long limit = stateMgr.offsetLimit(partition);
+        long limit = offsetLimits.getOrDefault(partition, Long.MAX_VALUE);
 
         long lastOffset = -1L;
         final List<ConsumerRecord<byte[], byte[]>> restoreRecords = new ArrayList<>(records.size());
         final List<ConsumerRecord<byte[], byte[]>> remainingRecords = new ArrayList<>();
 
+        // Check if we are unable to process one or more records due to an offset limit (e.g. when
+        // our partition is both a source and changelog). If we are limited then try to refresh
+        // the offset limit if possible.
+        if (!records.isEmpty() &&
+            records.get(records.size() - 1).offset() >= limit &&
+            updateableOffsetLimits.remove(partition)) {
```

Comment:
I know that using `remove` is more efficient, but it's somewhat header to extract to logic. Would be simpler to read if we use `contains`.

Reply:
No problem. Will switch to contains and do remove in the body.